File: easycontrol/server/src/main/java/top/saymzx/easycontrol/server/wrappers/WindowManager.java
/*
 * 本项目大量借鉴学习了开源投屏软件：Scrcpy，在此对该项目表示感谢
 */
package top.saymzx.easycontrol.server.wrappers;

import android.os.IInterface;
import android.view.Display;
import android.view.IRotationWatcher;

import java.lang.reflect.Method;

public final class WindowManager {
  private static IInterface manager;
  private static Class<?> CLASS;
  private static Method freezeRotationMethod = null;
  private static Method freezeDisplayRotationMethod = null;
  private static Method isRotationFrozenMethod = null;
  private static Method isDisplayRotationFrozenMethod = null;
  private static Method thawRotationMethod = null;
  private static Method thawDisplayRotationMethod = null;

  public static void init(IInterface m) {
    manager = m;
    CLASS = manager.getClass();
    try {
      freezeRotationMethod = manager.getClass().getMethod("freezeRotation", int.class);
      freezeDisplayRotationMethod = manager.getClass().getMethod("freezeDisplayRotation", int.class, int.class);
    } catch (Exception ignored) {
    }
    try {
      isRotationFrozenMethod = manager.getClass().getMethod("isRotationFrozen");
      isDisplayRotationFrozenMethod = manager.getClass().getMethod("isDisplayRotationFrozen", int.class);
    } catch (Exception ignored) {
    }
    try {
      thawRotationMethod = manager.getClass().getMethod("thawRotation");
      thawDisplayRotationMethod = manager.getClass().getMethod("thawDisplayRotation", int.class);
    } catch (Exception ignored) {
    }
  }

  public static void freezeRotation(int displayId, int rotation) {
    try {
      if (freezeDisplayRotationMethod == null) throw new Exception("no display");
      freezeDisplayRotationMethod.invoke(manager, displayId, rotation);
    } catch (Exception ignored) {
      try {
        if (freezeRotationMethod != null && displayId == Display.DEFAULT_DISPLAY) freezeRotationMethod.invoke(manager, rotation);
      } catch (Exception ignored1) {
      }
    }
  }

  public static boolean isRotationFrozen(int displayId) {
    try {
      if (isDisplayRotationFrozenMethod != null) throw new Exception("no display");
      return (boolean) isDisplayRotationFrozenMethod.invoke(manager, displayId);
    } catch (Exception ignored) {
      try {
        if (isRotationFrozenMethod != null && displayId == Display.DEFAULT_DISPLAY) return (boolean) isRotationFrozenMethod.invoke(manager);
        return false;
      } catch (Exception ignored1) {
        return false;
      }
    }
  }

  public static void thawRotation(int displayId) {
    try {
      if (thawDisplayRotationMethod != null) throw new Exception("no display");
      thawDisplayRotationMethod.invoke(manager, displayId);
    } catch (Exception ignored) {
      try {
        if (thawRotationMethod != null && displayId == Display.DEFAULT_DISPLAY) thawRotationMethod.invoke(manager);
      } catch (Exception ignored1) {
      }
    }
  }

  public static void registerRotationWatcher(IRotationWatcher rotationWatcher, int displayId) {
    try {
      try {
        CLASS.getMethod("watchRotation", IRotationWatcher.class, int.class).invoke(manager, rotationWatcher, displayId);
      } catch (NoSuchMethodException e) {
        CLASS.getMethod("watchRotation", IRotationWatcher.class).invoke(manager, rotationWatcher);
      }
    } catch (Exception ignored) {
    }
  }

}

File: scrcpy_android/server/src/main/java/top/saymzx/scrcpy/server/wrappers/WindowManager.kt
package top.saymzx.scrcpy.server.wrappers

import android.os.IInterface
import android.view.IRotationWatcher
import java.lang.reflect.InvocationTargetException

class WindowManager( // old version
  private val manager: IInterface
) {

  private var getRotationMethod = try {
    manager.javaClass.getMethod("getDefaultDisplayRotation")
  } catch (e: NoSuchMethodException) {
    manager.javaClass.getMethod("getRotation")
  }

  private var freezeRotationMethod =
    manager.javaClass.getMethod("freezeRotation", Int::class.javaPrimitiveType)

  private var isRotationFrozenMethod = manager.javaClass.getMethod("isRotationFrozen")

  private var thawRotationMethod = manager.javaClass.getMethod("thawRotation")

  val rotation: Int
    get() = try {
      getRotationMethod.invoke(manager) as Int
    } catch (e: InvocationTargetException) {
      0
    } catch (e: IllegalAccessException) {
      0
    } catch (e: NoSuchMethodException) {
      0
    }

  fun freezeRotation(rotation: Int) {
    try {
      freezeRotationMethod.invoke(manager, rotation)
    } catch (_: InvocationTargetException) {
    } catch (_: IllegalAccessException) {
    } catch (_: NoSuchMethodException) {
    }
  }

  val isRotationFrozen: Boolean
    get() = try {
      isRotationFrozenMethod.invoke(manager) as Boolean
    } catch (e: InvocationTargetException) {
      false
    } catch (e: IllegalAccessException) {
      false
    } catch (e: NoSuchMethodException) {
      false
    }

  fun thawRotation() {
    try {
      thawRotationMethod.invoke(manager)
    } catch (_: InvocationTargetException) {
    } catch (_: IllegalAccessException) {
    } catch (_: NoSuchMethodException) {
    }
  }

  fun registerRotationWatcher(rotationWatcher: IRotationWatcher?, displayId: Int) {
    try {
      val cls: Class<*> = manager.javaClass
      try {
        cls.getMethod("watchRotation", IRotationWatcher::class.java, Int::class.javaPrimitiveType)
          .invoke(
            manager, rotationWatcher, displayId
          )
      } catch (e: NoSuchMethodException) {
        cls.getMethod("watchRotation", IRotationWatcher::class.java)
          .invoke(manager, rotationWatcher)
      }
    } catch (_: Exception) {
    }
  }
}